### PR TITLE
Minimal encoding for Code-128

### PR DESCRIPTION
**Pull request by AlexGeller1**
Added CODE128_COMPACT encoding hint for activation.

### CODE REVIEW

**Comment by AlexGeller1** [comment]
I just saw that the diff for the method Code128Writer.chooseCode() looks huge but I didn't change it at all (0 lines). I just moved it into a Selector class so that I can choose between this implementation and the optimal implementation.

File: core/src/main/java/com/google/zxing/oned/Code128Writer.java [hunk 1]
```diff
@@ -179,26 +191,26 @@
...
-              default:
-                // CODE_CODE_C
-                if (position + 1 == length) {
-                  // this is the last character, but the encoding is C, which always encodes two characers
-                  throw new IllegalArgumentException("Bad number of characters for digit only encoding.");
-                }
-                patternIndex = Integer.parseInt(contents.substring(position, position + 2));
-                position++; // Also incremented below
-                break;
+            if (codeSet == CODE_CODE_A ||
```
Comment (srowen) [inline core/src/main/java/com/google/zxing/oned/Code128Writer.java:194]: Why remove the switch?

File: core/src/main/java/com/google/zxing/oned/Code128Writer.java [hunk 2]
```diff
@@ -282,66 +302,239 @@ private static CType findCType(CharSequence value, int start) {
...
-        if (c < ' ' || (oldCode == CODE_CODE_A && (c < '`' || (c >= ESCAPE_FNC_1 && c <= ESCAPE_FNC_4)))) {
-          // can continue in code A, encodes ASCII 0 to 95 or FNC1 to FNC4
-          return CODE_CODE_A;
+  }
+
+  /** 
+   * Encodes minimally using Divide-And-Conquer with Memoization
+   **/
+  private static class MinimalSelector implements Selector {
+    enum Charset { A, B, C, NONE };
```
Comment (srowen) [inline core/src/main/java/com/google/zxing/oned/Code128Writer.java:377]: private?

File: core/src/main/java/com/google/zxing/oned/Code128Writer.java (hunk 1 above)
Comment (AlexGeller1) [reply core/src/main/java/com/google/zxing/oned/Code128Writer.java:194]: Because the cases are are no longer only an integer

File: core/src/main/java/com/google/zxing/oned/Code128Writer.java [hunk 3]
```diff
@@ -179,26 +185,28 @@
...
-              default:
-                // CODE_CODE_C
-                if (position + 1 == length) {
-                  // this is the last character, but the encoding is C, which always encodes two characers
-                  throw new IllegalArgumentException("Bad number of characters for digit only encoding.");
-                }
-                patternIndex = Integer.parseInt(contents.substring(position, position + 2));
-                position++; // Also incremented below
-                break;
+            if ((codeSet == CODE_CODE_A &&
```
Comment (srowen) [inline core/src/main/java/com/google/zxing/oned/Code128Writer.java:188]: OK I see. I'm a little concerned about changing the exisitng code path; adding a second one isn't as big a deal. Are these changes necessary vs just having an alternative impl?

File: core/src/main/java/com/google/zxing/oned/Code128Writer.java [hunk 4]
```diff
@@ -282,66 +298,242 @@ private static CType findCType(CharSequence value, int start) {
...
-        char c = value.charAt(start);
-        if (c < ' ' || (oldCode == CODE_CODE_A && (c < '`' || (c >= ESCAPE_FNC_1 && c <= ESCAPE_FNC_4)))) {
-          // can continue in code A, encodes ASCII 0 to 95 or FNC1 to FNC4
-          return CODE_CODE_A;
+  }
+
+  /** 
+   * Encodes minimally using Divide-And-Conquer with Memoization
+   **/
+  private static class MinimalSelector implements Selector {
```
Comment (srowen) [inline core/src/main/java/com/google/zxing/oned/Code128Writer.java:372]: Like I'd kind of prefer the separate new code live in its own top level class, for clarity

Comment (AlexGeller1) [reply core/src/main/java/com/google/zxing/oned/Code128Writer.java:372]: Sure, no problem. This version was the option with the most code reuse in my opinion. By reimplementing only the lookahead function it shared the checks (checks for maximum length and check for illegal characters), the the building of the pattern, the weighted checksum accumulation, and the adding of the checksum and the stop pattern to the patterns.
To make it less invasive I see two alternative options.
1) Initially I wrote a version (also without introducing a new top level class) that has two code paths that compute the "pattern" ArrayList. That way it also shared the checks and the adding of the stop pattern (the checksum adding could also be shared but then the function would need to return two values (the ArrayList and the checksum) .
2) We rename Code128Writer.java to Code128FastWriter.java and add a new class Code128MinimalWriter.java. The checking code could be put in a separate function in Code128FastWriter.java and then be reused by calling it from Code128MinimalWriter too.
I think that both options are fine. The checksum computation and the adding of the stop pattern are so trivial that I don't think that it is worth the bother to try to share them so that I lean slightly more towards option 2. 
Which would you prefer or do you see yet another good way of doing it?

Comment (AlexGeller1) [reply core/src/main/java/com/google/zxing/oned/Code128Writer.java:372]: Update: On second thought, I am not so fond of option 2 after looking at the code some more. I propose to come up with a solution that leaves the exiting encode() function completely without modification and just dispatches between two versions of the function. This way the two functions can share the constants and the standard overrides of OneDimensionalCodeWriter. I will make a proposal and then we can discuss it from there.

Comment (AlexGeller1) [reply core/src/main/java/com/google/zxing/oned/Code128Writer.java:372]: Update: I created an new PR #1484 with a less intrusive version.
